src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -359,17 +360,32 @@ void remove_virtual_functionst::get_functions(
   }
 
   // iterate over all children, transitively
-  std::set<irep_idt> visited;
   get_child_functions_rec(
     class_id,
     root_function.symbol_expr,
     component_name,
     functions,
-    visited,
     resolve_function_call);
 
   if(root_function.symbol_expr!=symbol_exprt())
     functions.push_back(root_function);
+
+  // remove all classes from dispatch table where concrete call has been found
+  std::set<irep_idt> concrete_call;
+  for(const auto &entry : functions)
+  {
+    if(entry.symbol_expr != root_function.symbol_expr)
+      concrete_call.insert(entry.class_id);
+  }
+
+  std::remove_if(
+    functions.begin(),
+    functions.end(),
+    [&concrete_call, &root_function](const dispatch_table_entryt &entry) {
+      return
+        concrete_call.find(entry.class_id) != concrete_call.end() &&
+        entry.symbol_expr == root_function.symbol_expr;
+    });
```

Comment:
I maintain that I don't understand why this code has to be as complicated.

Reply:
I don't know if it is complicated but it seems to do some unnecessary sorting.
Maybe pull out this code in a function which makes the intent clear, something like `sort_java_lang_object_at_the_end(dispatch_table_entries &)`.
Then it will be easier to see how it can be improved.